app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,51 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user,         only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
+
+  def new
+  end
+
+  def edit
+  end
+
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:notice] = "Email sent with reset instructions"
+      redirect_to root_url
+    else
+      flash.now[:error] = "Email address not found"
+      render "new"
+    end
+  end
+
+  def update
+    if @user.update_attributes(user_params)
+      log_in @user
+      flash[:success] = "Password has been reset"
+      redirect_to @user
+    else
+      render "edit"
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+
+  def get_user
+    @user = User.find_by(email: params[:email])
+    redirect_to root_url if @user.nil?
+  end
+
+  def check_expiration
```

Comment:
I think that in addition to checking the token expiration, the user needs to provide the token so we can confirm they received the email. The way it works is you generate a password reset token and then you store a version of the token hashed (with BCrypt). The email that we sent to the user has the un-hashed version of the token, as a parameter to the `update` endpoint. When you get to the `update` action you hash the token from say `params[:token]` and then look up the user using the hashed token. If you have one then we check confirmation and proceed, if not then the token is invalid and we can stop.

We can talk more about this flow in our next session :smile: